ipysurfer/_utils/funcs.py

```diff
@@ -0,0 +1,203 @@
+import os
+import logging
+
+import numpy as np
+from scipy import sparse
+
+
+logger = logging.getLogger('ipysurfer')
+
+
+def _check_units(units):
+    if units not in ('m', 'mm'):
+        raise ValueError('Units must be "m" or "mm", got %r' % (units,))
+    return units
+
+
+def _compute_normals(rr, tris):
+    """Efficiently compute vertex normals for triangulated surface."""
+    # first, compute triangle normals
+    r1 = rr[tris[:, 0], :]
+    r2 = rr[tris[:, 1], :]
+    r3 = rr[tris[:, 2], :]
+    tri_nn = _fast_cross_3d((r2 - r1), (r3 - r1))
+
+    #   Triangle normals and areas
+    size = np.sqrt(np.sum(tri_nn * tri_nn, axis=1))
+    zidx = np.where(size == 0)[0]
+    size[zidx] = 1.0  # prevent ugly divide-by-zero
+    tri_nn /= size[:, np.newaxis]
+
+    npts = len(rr)
+
+    # the following code replaces this, but is faster (vectorized):
+    #
+    # for p, verts in enumerate(tris):
+    #     nn[verts, :] += tri_nn[p, :]
+    #
+    nn = np.zeros((npts, 3))
+    for verts in tris.T:  # note this only loops 3x (number of verts per tri)
+        for idx in range(3):  # x, y, z
+            nn[:, idx] += np.bincount(verts, tri_nn[:, idx], minlength=npts)
+    size = np.sqrt(np.sum(nn * nn, axis=1))
+    size[size == 0] = 1.0  # prevent ugly divide-by-zero
+    nn /= size[:, np.newaxis]
+    return nn
+
+
+def _fast_cross_3d(x, y):
+    """Compute cross product between list of 3D vectors.
+
+    Much faster than np.cross() when the number of cross products
```

Comment:
Are these copied over from MNE or custom for this repo?

Reply:
I have used `PySurfer` source code for these. As far as I understand, `ipysurfer` should be a stand-alone package, so I was trying not to depend on `PySurfer` or create a circular dependency with `MNE`.

Reply:
in general I'm fine with this package depending on pysurfer and using some of its utility functions, unless that would make "mayavi" a dependency of ipysurfer. @larsoner do you know if that's a reason not to depend on pysurfer?

Reply:
I was avoiding dependency on `PySurfer` due to possible `mayavi` involvement.